Add runbook for adding and removing storage nodes (#102)
appuio/openshift4-docs: docs/modules/ROOT/pages/how-tos/exoscale/remove_storage_node.adoc

= Remove a storage node

:kubectl_extra_args: --as=cluster-admin
:delabel_app_nodes: yes
:argo_app: rook-ceph
:osd-node: ${NODE_TO_REMOVE}
:mon-node: ${NODE_TO_REMOVE}
:mon-operation: remove
:delete-node: ${NODE_TO_REMOVE}
:delete-node-manually: no

[abstract]
--
Steps to remove a storage node of an OpenShift 4 cluster on https://www.exoscale.com[Exoscale].
--

== Starting situation

* You already have a OpenShift 4 cluster on Exoscale
* You have admin-level access to the cluster
* You want to remove an existing storage node in the cluster

== Prerequisites

include::partial$exoscale/prerequisites.adoc[]

== Prepare local environment

include::partial$exoscale/setup-local-env.adoc[]

== Set alert silence

:duration: +60 minutes
include::partial$create-alertmanager-silence.adoc[]

== Update Cluster Config

. Update cluster config. 
+
[source,bash]
----
pushd "inventory/classes/${TENANT_ID}/"

yq eval -i ".parameters.openshift4_terraform.terraform_variables.storage_count -= 1" \
  ${CLUSTER_ID}.yml

yq eval -i ".parameters.rook_ceph.ceph_cluster.node_count -= 1" \
  ${CLUSTER_ID}.yml
----
+
[NOTE]
====

> Ceph can't scale below 3 storage nodes, which is the default number of nodes.

Reviewer: ```suggestion
Ceph can't scale to less than 3 storage nodes, which is the default number of nodes.
```

> Please check that this update doesn't reduce the number of storage nodes below 3, before continuing.

Reviewer: ```suggestion
Please ensure that this update doesn't reduce the number of storage nodes to less than 3 before continuing.
```

====

. Review and commit
+
[source,bash]
----

# Have a look at the file ${CLUSTER_ID}.yml.

git commit -a -m "Remove storage node to cluster ${CLUSTER_ID}"
git push

popd
----

. Compile and push cluster catalog 
+
[source,bash]
----
commodore catalog compile ${CLUSTER_ID} --push -i
----

== Prepare Terraform environment

include::partial$exoscale/configure-terraform-secrets.adoc[]

include::partial$setup_terraform.adoc[]

== Remove Node
* Find the node you want to remove.
It has to be the one with the highest terraform index.
+
[source,bash]
----
# Grab JSON copy of current Terraform state
terraform state pull > .tfstate.json

node_count=$(jq  -r \
  '.resources[] |
   select(.module=="module.cluster.module.storage" and .type=="exoscale_compute") |
   .instances | length' \
   .tfstate.json)

> # Make sure the number of nodes is conistent with what we configured earlier

Reviewer: ```suggestion
# Verify that the number of nodes is one more than we configured earlier.
```

* `vault` https://www.vaultproject.io/docs/commands[Vault CLI]
* `commodore`, see https://syn.tools/commodore/running-commodore.html[Running Commodore]
* `jq`
* `yq` https://mikefarah.gitbook.io/yq[yq YAML processor] (version 4 or higher)

== Prepare local environment

. Create local directory to work in
+
[TIP]
====
We strongly recommend creating an empty directory, unless you already have a work directory for the cluster you're about to work on.
This guide will run Commodore in the directory created in this step.
====
+
[source,bash]
----
export WORK_DIR=/path/to/work/dir
mkdir -p "${WORK_DIR}"
pushd "${WORK_DIR}"
----

. Configure API access
+
include::partial$exoscale/environment-vars.adoc[]
+
include::partial$vshn-input.adoc[]

. Get required tokens from Vault
+
include::partial$connect-to-vault.adoc[]
+
include::partial$get-hieradata-token-from-vault.adoc[]

. Compile the catalog for the cluster.
Having the catalog available locally enables us to run Terraform for the cluster to make any required changes.
+
[source,bash]
----
commodore catalog compile "${CLUSTER_ID}"
----

== Set alert silence

:duration: +60 minutes

> include::partial$create-alertmanager-silence.adoc[]

Reviewer: From what I can see, this whole section is identical between this how-to, adding a node and replacing a node. Probably makes sense to move this to a partial, as any changes here will need to be reflected in all three howtos anyway.

== Update Cluster Config

. Update cluster config. 
+
[source,bash]
----
pushd "inventory/classes/${TENANT_ID}/"

yq eval -i ".parameters.openshift4_terraform.terraform_variables.storage_count -= 1" \
  ${CLUSTER_ID}.yml

yq eval -i ".parameters.rook_ceph.ceph_cluster.node_count -= 1" \
  ${CLUSTER_ID}.yml

----
+
[NOTE]
====
Ceph can't scale below 3 storage nodes, which is the default number of nodes.
Please check that this update doesn't reduce the number of storage nodes below 3, before continuing.
====

. Review and commit
+
[source,bash]
----

# Have a look at the file ${CLUSTER_ID}.yml.

git commit -a -m "Remove storage node to cluster ${CLUSTER_ID}"
git push

popd
----

. Compile and push cluster catalog 
+
[source,bash]
----
commodore catalog compile ${CLUSTER_ID} --push -i
----

== Prepare Terraform environment

include::partial$exoscale/configure-terraform-secrets.adoc[]

include::partial$setup_terraform.adoc[]

== Remove Node

> Make a note of the node you want to remove

Reviewer: Maybe make this a list item?

[source,bash]
----
export NODE_TO_REMOVE=storage-XXXX
----

=== Remove old OSD
. Make sure ArgoCD ran and reduced the target number of OSDs
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-cluster \
  get cephcluster cluster -o jsonpath='{.spec.storage.storageClassDeviceSets[0].count}'
----

> :osd-node: ${NODE_TO_REMOVE}

Reviewer: Define at top of the page

+
[source,bash]
----
pvc_name=$(kubectl --as=cluster-admin -n syn-rook-ceph-cluster get deploy \
  "rook-ceph-osd-${OSD_ID}" -ojsonpath='{.metadata.labels.ceph\.rook\.io/pvc}')
pv_name=$(kubectl --as=cluster-admin -n syn-rook-ceph-cluster get pvc \
  "${pvc_name}" -o jsonpath='{.spec.volumeName}')
----

. Check if the OSD deployment needs to be deleted, and delete it if necessary
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-cluster get deploy \
  -l failure-domain="${NODE_TO_REMOVE}"
# Run this command if the previous command lists a deployment
kubectl --as=cluster-admin -n syn-rook-ceph-cluster delete deploy \
  -l failure-domain="${NODE_TO_REMOVE}"
----

. Clean up PVC and prepare job of the old OSD if necessary
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-cluster delete job \
  -l ceph.rook.io/pvc="${pvc_name}"
kubectl --as=cluster-admin -n syn-rook-ceph-cluster delete pvc "${pvc_name}"
----

. Clean up PVC encryption secret
+
[source,bash]
----
kubectl --as=cluster-admin -n syn-rook-ceph-cluster delete secret -l pvc_name="${pvc_name}"

> ----

Reviewer: Either add this to the `storage-ceph-remove-osd.adoc` partial (including scaling the operator step (conditional on another attribute) which is omitted here), or create another partial for this part, as it seems identical to the steps for replacing a node.

=== Remove the old MON

:mon_node: ${NODE_TO_REMOVE}
include::partial$storage-ceph-remove-mon.adoc[]

=== Remove VM

. Find Terraform resource index of the node to replace
+
[source,bash]
----
# Grab JSON copy of current Terraform state
terraform state pull > .tfstate.json
node_index=$(jq --arg storage_node "${NODE_TO_REMOVE}" -r \
  '.resources[] |
   select(.module=="module.cluster.module.storage" and .type=="random_id") |
   .instances[] |
   select(.attributes.hex==$storage_node) |
   .index_key' \
  .tfstate.json)
----

. Verify that resource index is correct
+
[source,bash]
----
jq --arg index "${node_index}" -r \
  '.resources[] |
   select(.module=="module.cluster.module.storage" and .type=="exoscale_compute") |
   .instances[$index|tonumber] |
   .attributes.hostname' \
   .tfstate.json
----

. Remove node ID and node resource for node that we want to remove from the Terraform state
+
[source,bash]
----
terraform state rm "module.cluster.module.storage.random_id.node_id[$node_index]"
terraform state rm "module.cluster.module.storage.exoscale_compute.nodes[$node_index]"
----

> :delete-node: ${NODE_TO_REMOVE}

Reviewer: Defining the page attribute between list items breaks the list flow and the included items from the partial are numbered starting from 1. Generally, it's safer to define attributes at the top.

> include::partial$exoscale/delete-node.adoc[]

Reviewer: This will completely break subsequent runs of `terraform apply` if any storage node except the one with the highest index in the Terraform state is removed.

I'd suggest extracting the node to remove from the Terraform state (highest ID in the state) and adjust the how-to to remove that node. Deleting the VM on Exoscale then becomes

```
terraform apply
```

. Clean up localstorage PV of decommissioned node
+
[source,bash]
----
kubectl --as=cluster-admin delete pv "${pv_name}"

> ----

Reviewer: Pretty sure this can go into `delete-node.adoc`
